# Assignment 3 Ruisi_Gu -- Sub-Notebook Two

## Part Three -- Fit in Models and Comparison

This is the beginning of the second notebook. In this part, I am using H2O to do the machine learning work, which is *H2OAutoML*, also I will generate and compare these models to get the best result. (What need to illustrate is that work which is done in this notebook is based on the dataset generated from the first notebook.)

#### Initialize H2O

In [1]:
import h2o
from h2o.automl import H2OAutoML
import random, os, sys
from datetime import datetime
import pandas as pd
import logging
import csv
import optparse
import time
import json
from distutils.util import strtobool
import psutil
import numpy as np

In [2]:
pct_memory=8
virtual_memory=psutil.virtual_memory()
min_mem_size=int(round(int(pct_memory*virtual_memory.available)/1073741824,0))
print(min_mem_size)

12


In [3]:
h2o.init(strict_version_check=False,max_mem_size=min_mem_size)

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
  Java Version: java version "1.8.0_191"; Java(TM) SE Runtime Environment (build 1.8.0_191-b12); Java HotSpot(TM) 64-Bit Server VM (build 25.191-b12, mixed mode)
  Starting server from /Users/guruisi/anaconda3/lib/python3.7/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /var/folders/25/079nl8ps1d17m_b9mm9cwnhm0000gn/T/tmp0pgzh4ov
  JVM stdout: /var/folders/25/079nl8ps1d17m_b9mm9cwnhm0000gn/T/tmp0pgzh4ov/h2o_guruisi_started_from_python.out
  JVM stderr: /var/folders/25/079nl8ps1d17m_b9mm9cwnhm0000gn/T/tmp0pgzh4ov/h2o_guruisi_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321... successful.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         03 secs
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.22.1.5
H2O cluster version age:    29 days
H2O cluster name:           H2O_from_python_guruisi_92tcg0
H2O cluster total nodes:    1
H2O cluster free memory:    10.67 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.7.1 final
--------------------------  ---------------------------------------------------

In [38]:
# train_data_path = "assignment3/data/train.tsv"
# test_data_path = "assignment3/data/test.tsv"
train_data_path = "assignment3/train_submission.csv"
test_data_path = "assignment3/submission06.csv"

# Load data into H2O
df_train = h2o.import_file(train_data_path)
df_test = h2o.import_file(test_data_path)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [5]:
df_train.head()

PhraseId SentenceId Phrase Sentiment SentimentM1 
 
 
 1 1 A series of escapades demonstrating the adage that what is good for the goose is also good for the gander , some of which occasionally amuses but none of which amounts to much of a story . 1 3 
 2 1 A series of escapades demonstrating the adage that what is good for the goose 2 2 
 3 1 A series 2 2 
 4 1 A 2 2 
 5 1 series 2 2 
 6 1 of escapades demonstrating the adage that what is good for the goose 2 2 
 7 1 of 2 2 
 8 1 escapades demonstrating the adage that what is good for the goose 2 2 
 9 1 escapades 2 2 
 10 1 demonstrating the adage that what is good for the goose 2 2

In [39]:
df_test.head()

PhraseId SentimentM1 
 
 
 156061 3 
 156062 3 
 156063 2 
 156064 3 
 156065 3 
 156066 3 
 156067 3 
 156068 3 
 156069 3 
 156070 2

In our case, although Sentiment labels range from 0 to 4, Sentiment is not represent values 0,1,2,3,4. Thus, we need to convert the numbers into classifiers in case H2O consider the data as linear regression.

In [7]:
df_train['Sentiment'] = df_train['Sentiment'].asfactor()

In [8]:
df_train.head()

PhraseId SentenceId Phrase Sentiment SentimentM1 
 
 
 1 1 A series of escapades demonstrating the adage that what is good for the goose is also good for the gander , some of which occasionally amuses but none of which amounts to much of a story . 1 3 
 2 1 A series of escapades demonstrating the adage that what is good for the goose 2 2 
 3 1 A series 2 2 
 4 1 A 2 2 
 5 1 series 2 2 
 6 1 of escapades demonstrating the adage that what is good for the goose 2 2 
 7 1 of 2 2 
 8 1 escapades demonstrating the adage that what is good for the goose 2 2 
 9 1 escapades 2 2 
 10 1 demonstrating the adage that what is good for the goose 2 2

#### Processing data by using H2O AutoML. 
Set runtime 15000 seconds in order to get a better model. Then choose the best model from leaderboard as below.

In [9]:
def get_independent_variables(df, targ):
    C = [name for name in df.columns if name != targ]
    # determine column types
    ints, reals, enums = [], [], []
    for key, val in df.types.items():
        if key in C:
            if val == 'enum':
                enums.append(key)
            elif val == 'int':
                ints.append(key)            
            else: 
                reals.append(key)    
    x=ints+enums+reals
    return x

In [10]:
target='Sentiment'
y = target

In [11]:
X=get_independent_variables(df_train, target) 
print(X)

['PhraseId', 'SentenceId', 'SentimentM1', 'Phrase']


In [12]:
run_time=1500

In [13]:
review_train, review_test = df_train.split_frame([0.1])
review_train.shape

(15610, 5)

In [14]:
# Set up AutoML
aml = H2OAutoML(max_runtime_secs=run_time)

In [15]:
aml.train(x=X,y=y,training_frame=df_train)

AutoML progress: |████████████████████████████████████████████████████████| 100%


In [16]:
print(aml.leaderboard)

model_id mean_per_class_error logloss rmse mse 
 
 
 XGBoost_1_AutoML_20190403_185203 0.518557 1.05243 0.614541 0.377661 
 GLM_grid_1_AutoML_20190403_185203_model_1 0.519231 1.03804 0.59256 0.351127 
 GBM_1_AutoML_20190403_185203 0.521965 1.00825 0.588014 0.34576 
 XGBoost_3_AutoML_20190403_185203 0.521969 1.11105 0.645938 0.417236 
 XGBoost_2_AutoML_20190403_185203 0.522008 1.11353 0.645934 0.41723 
 GBM_2_AutoML_20190403_185203 0.52578 1.03058 0.602379 0.362861 
 GBM_3_AutoML_20190403_185203 0.526781 1.01908 0.597172 0.356614 
 DRF_1_AutoML_20190403_185203 0.527867 0.94295 0.557172 0.31044 
 StackedEnsemble_AllModels_AutoML_20190403_185203 0.528813 0.908494 0.552658 0.305431 
 StackedEnsemble_BestOfFamily_AutoML_20190403_185203 0.530084 0.908995 0.552754 0.305537

Based on *mean_per_class_error* and *logloss*, I've selected two models that perform well. One is XGBoost model, which run the first place as we can see from the leaderboard. The other one is GBM (Gradient Boosting Model) which has the least logloss.<br><br>
I will name the **XGBoost** Model as **mod_best**, at the same time, **GBM** Model as **model**

In [17]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [18]:
mod_best

Model Details
H2OXGBoostEstimator :  XGBoost
Model Key:  XGBoost_1_AutoML_20190403_185203


ModelMetricsMultinomial: xgboost
** Reported on train data. **

MSE: 0.3782131116029184
RMSE: 0.6149903345605672
LogLoss: 1.0533676846424738
Mean Per-Class Error: 0.5222057126000716
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2       3      4     Error     Rate
----  -----  ------  -----  ----  --------  ----------------
2183  2773   1914    176    26    0.691318  4,889 / 7,072
1218  10746  13981   1234   94    0.605984  16,527 / 27,273
381   4641   68415   5765   380   0.140321  11,167 / 79,582
69    1135   14567   15289  1867  0.53567   17,638 / 32,927
7     159    1694    4011   3335  0.637736  5,871 / 9,206
3858  19454  100571  26475  5702  0.359426  56,092 / 156,060

Top-5 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.640574
2    0.826426
3    0.958868
4    0.988242
5    1


ModelMetricsMultinomial: xgboost
** Reported on cross-validation data. **

MSE: 0.37766073429299074
RMSE: 0.6145410761641493
LogLoss: 1.0524307259660532
Mean Per-Class Error: 0.5185567303270829
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2       3      4     Error     Rate
----  -----  ------  -----  ----  --------  ----------------
2363  2593   1914    176    26    0.665865  4,709 / 7,072
1403  10561  13979   1234   96    0.612767  16,712 / 27,273
446   4576   68412   5766   382   0.140358  11,170 / 79,582
78    1126   14569   15287  1867  0.535731  17,640 / 32,927
7     159    1694    4014   3332  0.638062  5,874 / 9,206
4297  19015  100568  26477  5703  0.359509  56,105 / 156,060

Top-5 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.640491
2    0.824817
3    0.958375
4    0.986621
5    1

Cross-Validation Metrics Summary: 


mean      sd          cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
-----------------------  --------  ----------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.640491  0.00159134  0.640331      0.642766      0.636262      0.64158       0.641516
err                      0.359509  0.00159134  0.359669      0.357234      0.363738      0.35842       0.358484
err_count                11221     49.6689     11226         11150         11353         11187         11189
logloss                  1.05243   0.00361426  1.04841       1.05677       1.06008       1.0468        1.0501
max_per_class_error      0.665818  0.00260722  0.666667      0.658747      0.668772      0.666189      0.668716
mean_per_class_accuracy  0.481439  0.00248362  0.483136      0.485609      0.476109      0.483706      0.478635
mean_per_class_error     0.518561  0.00248362  0.516864      0.51439       0.523891      0.516294      0.521365
mse                      0.377661  0.00232513  0.37524       0.381007      0.38228       0.374508      0.375268
r2                       0.527272  0.003781    0.532164      0.523757      0.518732      0.528843      0.532864
rmse                     0.614535  0.00189001  0.612569      0.617258      0.618288      0.61197       0.612591

Scoring History: 


timestamp            duration          number_of_trees    training_rmse    training_logloss    training_classification_error
--  -------------------  ----------------  -----------------  ---------------  ------------------  -------------------------------
    2019-04-03 19:07:35  1 min 58.716 sec  0                  0.8              1.60944             0.789011
    2019-04-03 19:07:36  1 min 59.939 sec  5                  0.767999         1.4678              0.40585
    2019-04-03 19:07:37  2 min  1.001 sec  10                 0.738676         1.36099             0.359958
    2019-04-03 19:07:38  2 min  2.007 sec  15                 0.713635         1.28245             0.359413
    2019-04-03 19:07:39  2 min  3.076 sec  20                 0.690541         1.21844             0.359407
    2019-04-03 19:07:41  2 min  4.204 sec  25                 0.673644         1.17605             0.359426
    2019-04-03 19:07:42  2 min  5.407 sec  30                 0.660216         1.14502             0.359214
    2019-04-03 19:07:43  2 min  6.636 sec  35                 0.648157         1.11848             0.359432
    2019-04-03 19:07:44  2 min  7.938 sec  40                 0.63794          1.0974              0.35958
    2019-04-03 19:07:46  2 min  9.388 sec  45                 0.63031          1.08229             0.359413
    2019-04-03 19:07:47  2 min 10.800 sec  50                 0.624345         1.07089             0.359458
    2019-04-03 19:07:49  2 min 12.510 sec  55                 0.618135         1.0589              0.359458
    2019-04-03 19:07:50  2 min 14.051 sec  58                 0.61499          1.05337             0.359426

Variable Importances: 


variable     relative_importance    scaled_importance    percentage
-----------  ---------------------  -------------------  ------------
SentimentM1  595630                 1                    0.974636
PhraseId     8686.37                0.0145835            0.0142136
SentenceId   6814.05                0.0114401            0.0111499

In [19]:
mod_best.algo

'xgboost'

In [20]:
model = h2o.get_model(aml.leaderboard[5,'model_id'])
model.algo

'gbm'

In [21]:
model

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_2_AutoML_20190403_185203


ModelMetricsMultinomial: gbm
** Reported on train data. **

MSE: 0.3602525771288419
RMSE: 0.600210444035125
LogLoss: 1.023761984731218
Mean Per-Class Error: 0.5212745950931236
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2       3      4     Error     Rate
----  -----  ------  -----  ----  --------  ----------------
2264  2693   1914    177    24    0.679864  4,808 / 7,072
1243  10727  13995   1223   85    0.606681  16,546 / 27,273
377   4626   68477   5762   340   0.139542  11,105 / 79,582
69    1133   14577   15438  1710  0.531145  17,489 / 32,927
7     157    1698    4114   3230  0.649142  5,976 / 9,206
3960  19336  100661  26714  5389  0.358349  55,924 / 156,060

Top-5 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.641651
2    0.833019
3    0.958458
4    0.986697
5    1


ModelMetricsMultinomial: gbm
** Reported on cross-validation data. **

MSE: 0.362860776738782
RMSE: 0.6023792632044881
LogLoss: 1.0305772465589829
Mean Per-Class Error: 0.5257796504218959
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2       3      4     Error     Rate
----  -----  ------  -----  ----  --------  ----------------
2184  2768   1918    178    24    0.691176  4,888 / 7,072
1278  10683  14002   1225   85    0.608294  16,590 / 27,273
376   4642   68439   5789   336   0.140019  11,143 / 79,582
67    1137   14587   15388  1748  0.532663  17,539 / 32,927
7     158    1696    4185   3160  0.656746  6,046 / 9,206
3912  19388  100642  26765  5353  0.360156  56,206 / 156,060

Top-5 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.639844
2    0.829399
3    0.95785
4    0.986275
5    1

Cross-Validation Metrics Summary: 


mean      sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
-----------------------  --------  -----------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.639844  0.00167631   0.639337      0.64174       0.635461      0.641869      0.640811
err                      0.360156  0.00167631   0.360663      0.35826       0.364539      0.358131      0.359189
err_count                11241.2   52.3209      11257         11182         11378         11178         11211
logloss                  1.03058   0.00158138   1.03053       1.0299        1.03413       1.02717       1.03115
max_per_class_error      0.691106  0.00572898   0.701001      0.681065      0.696842      0.681948      0.694672
mean_per_class_accuracy  0.474224  0.0027388    0.474029      0.478374      0.468528      0.478543      0.471643
mean_per_class_error     0.525776  0.0027388    0.525971      0.521626      0.531472      0.521457      0.528357
mse                      0.362861  0.000435392  0.363152      0.362721      0.363794      0.361911      0.362726
r2                       0.545804  0.0015975    0.547235      0.546615      0.542005      0.544691      0.548477
rmse                     0.602379  0.000361397  0.602621      0.602263      0.603153      0.60159       0.602267

Scoring History: 


timestamp            duration          number_of_trees    training_rmse    training_logloss    training_classification_error
--  -------------------  ----------------  -----------------  ---------------  ------------------  -------------------------------
    2019-04-03 19:15:11  1 min  1.491 sec  0                  0.8              1.60944             0.659759
    2019-04-03 19:15:13  1 min  3.092 sec  5                  0.704969         1.25785             0.358772
    2019-04-03 19:15:15  1 min  4.923 sec  10                 0.647013         1.11435             0.358843
    2019-04-03 19:15:17  1 min  6.564 sec  15                 0.620209         1.05964             0.358503
    2019-04-03 19:15:18  1 min  7.697 sec  20                 0.610601         1.0429              0.358452
    2019-04-03 19:15:19  1 min  8.675 sec  25                 0.602888         1.02894             0.358253
    2019-04-03 19:15:19  1 min  9.176 sec  30                 0.60021          1.02376             0.358349

Variable Importances: 


variable     relative_importance    scaled_importance    percentage
-----------  ---------------------  -------------------  ------------
SentimentM1  78767.1                1                    0.945113
SentenceId   2428.47                0.030831             0.0291388
PhraseId     2145.88                0.0272434            0.0257481

In [22]:
# Find the 80th quantile of PhraseId in the dataset
Phrase_split = df_train["PhraseId"].quantile(prob = [0.8])[1]
df_train["Train"] = (df_train["PhraseId"] < Phrase_split).ifelse("Yes", "No")

In [23]:
df_train1 = df_train[df_train["Train"] == "Yes"]
df_test1 = df_train[df_train["Train"] == "No"]

In [24]:
df_train

PhraseId SentenceId Phrase Sentiment SentimentM1 Train 
 
 
 1 1 A series of escapades demonstrating the adage that what is good for the goose is also good for the gander , some of which occasionally amuses but none of which amounts to much of a story . 1 3 Yes 
 2 1 A series of escapades demonstrating the adage that what is good for the goose 2 2 Yes 
 3 1 A series 2 2 Yes 
 4 1 A 2 2 Yes 
 5 1 series 2 2 Yes 
 6 1 of escapades demonstrating the adage that what is good for the goose 2 2 Yes 
 7 1 of 2 2 Yes 
 8 1 escapades demonstrating the adage that what is good for the goose 2 2 Yes 
 9 1 escapades 2 2 Yes 
 10 1 demonstrating the adage that what is good for the goose 2 2 Yes

In [25]:
df_train1

PhraseId SentenceId Phrase Sentiment SentimentM1 Train 
 
 
 1 1 A series of escapades demonstrating the adage that what is good for the goose is also good for the gander , some of which occasionally amuses but none of which amounts to much of a story . 1 3 Yes 
 2 1 A series of escapades demonstrating the adage that what is good for the goose 2 2 Yes 
 3 1 A series 2 2 Yes 
 4 1 A 2 2 Yes 
 5 1 series 2 2 Yes 
 6 1 of escapades demonstrating the adage that what is good for the goose 2 2 Yes 
 7 1 of 2 2 Yes 
 8 1 escapades demonstrating the adage that what is good for the goose 2 2 Yes 
 9 1 escapades 2 2 Yes 
 10 1 demonstrating the adage that what is good for the goose 2 2 Yes

In [26]:
df_test1

In [40]:
X_train = df_train['Phrase']
y_train = df_train['Sentiment']
X_test = df_test['Phrase']

In [29]:
from h2o.estimators import H2OGradientBoostingEstimator
predictors = ['PhraseId', 'SentenceId', 'SentimentM1', 'Phrase']
response = 'Sentiment'

gbm_embeddings = H2OGradientBoostingEstimator(model_id = "gbm_embeddings.hex")
gbm_embeddings.train(x = predictors, y = response, 
                   training_frame = df_train1, validation_frame = df_test1
                  )

gbm Model Build progress: |███████████████████████████████████████████████| 100%


In [30]:
from h2o.estimators import H2OXGBoostEstimator
predictors = ['PhraseId', 'SentenceId', 'SentimentM1', 'Phrase']
response = 'Sentiment'
param = {
      "ntrees" : 100
    , "max_depth" : 10
    , "learn_rate" : 0.02
    , "sample_rate" : 0.7
    , "col_sample_rate_per_tree" : 0.9
    , "min_rows" : 5
    , "seed": 4241
    , "score_tree_interval": 100
}

model_xgboost = H2OXGBoostEstimator(**param)
model_xgboost.train(x = predictors, y = response, 
                   training_frame = df_train1, validation_frame = df_test1)

xgboost Model Build progress: |███████████████████████████████████████████| 100%


In [32]:
cf_xgboost=model_xgboost.confusion_matrix(df_train)
print(cf_xgboost)

Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2      3      4     Error     Rate
----  -----  -----  -----  ----  --------  ----------------
1886  3055   1506   600    25    0.733314  5,186 / 7,072
985   10955  11036  4209   88    0.598321  16,318 / 27,273
291   4661   55543  18757  330   0.302066  24,039 / 79,582
50    1138   11688  18518  1533  0.437604  14,409 / 32,927
6     159    1350   4524   3167  0.655985  6,039 / 9,206
3218  19968  81123  46608  5143  0.422857  65,991 / 156,060

In [34]:
cf=gbm_embeddings.confusion_matrix(df_train)
print(cf)
# print("With Embeddings AUC: " + str(round(gbm_embeddings.auc(valid = True), 3)))

Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2      3      4     Error     Rate
----  -----  -----  -----  ----  --------  ----------------
2346  2611   1490   599    26    0.668269  4,726 / 7,072
1346  10622  10992  4220   93    0.610531  16,651 / 27,273
417   4587   55395  18816  367   0.303926  24,187 / 79,582
76    1127   11643  18258  1823  0.445501  14,669 / 32,927
7     159    1342   4381   3317  0.639692  5,889 / 9,206
4192  19106  80862  46274  5626  0.423696  66,122 / 156,060

In [35]:
# gbm_embeddings.auc(valid = True)
cf=gbm_embeddings.confusion_matrix(df_train)

In [36]:
cf

Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



0     1      2      3      4     Error     Rate
----  -----  -----  -----  ----  --------  ----------------
2346  2611   1490   599    26    0.668269  4,726 / 7,072
1346  10622  10992  4220   93    0.610531  16,651 / 27,273
417   4587   55395  18816  367   0.303926  24,187 / 79,582
76    1127   11643  18258  1823  0.445501  14,669 / 32,927
7     159    1342   4381   3317  0.639692  5,889 / 9,206
4192  19106  80862  46274  5626  0.423696  66,122 / 156,060

In [ ]:
# cf.auc

In [55]:
# final_result1 = gbm_embeddings.predict(X_test)
predict = model_xgboost.predict(df_test)

xgboost prediction progress: |████████████████████████████████████████████| 100%


In [56]:
predict

predict p0 p1 p2 p3 p4 
 
 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 2 0.0805012 0.115926 0.398743 0.263018 0.141812 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 3 0.0610714 0.0833332 0.101097 0.549312 0.205186 
 2 0.0805012 0.115926 0.398743 0.263018 0.141812

In [57]:
result_Sentiment = predict['predict']

In [58]:
result_Sentiment

predict 
 
 
 3 
 3 
 2 
 3 
 3 
 3 
 3 
 3 
 3 
 2

In [59]:
type(predict)

h2o.frame.H2OFrame

In [60]:

result_data_path = "assignment3/data/test.tsv"

# Load data into H2O

df_result = h2o.import_file(result_data_path, sep="\t")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [61]:
df_result

PhraseId SentenceId Phrase 
 
 
 156061 8545 An intermittently pleasing but mostly routine effort . 
 156062 8545 An intermittently pleasing but mostly routine effort 
 156063 8545 An 
 156064 8545 intermittently pleasing but mostly routine effort 
 156065 8545 intermittently pleasing but mostly routine 
 156066 8545 intermittently pleasing but 
 156067 8545 intermittently pleasing 
 156068 8545 intermittently 
 156069 8545 pleasing 
 156070 8545 but

In [62]:
df_result = df_result.cbind(result_Sentiment)

In [63]:
df_result

PhraseId SentenceId Phrase predict 
 
 
 156061 8545 An intermittently pleasing but mostly routine effort . 3 
 156062 8545 An intermittently pleasing but mostly routine effort 3 
 156063 8545 An 2 
 156064 8545 intermittently pleasing but mostly routine effort 3 
 156065 8545 intermittently pleasing but mostly routine 3 
 156066 8545 intermittently pleasing but 3 
 156067 8545 intermittently pleasing 3 
 156068 8545 intermittently 3 
 156069 8545 pleasing 3 
 156070 8545 but 2

In [64]:
Submission = df_result.drop(['SentenceId','Phrase'])

In [65]:
Submission

PhraseId predict 
 
 
 156061 3 
 156062 3 
 156063 2 
 156064 3 
 156065 3 
 156066 3 
 156067 3 
 156068 3 
 156069 3 
 156070 2

In [66]:
Submission.set_names(['PhraseId','Sentiment'])

PhraseId Sentiment 
 
 
 156061 3 
 156062 3 
 156063 2 
 156064 3 
 156065 3 
 156066 3 
 156067 3 
 156068 3 
 156069 3 
 156070 2

In [68]:
h2o.export_file(Submission, 'Submission_plza.csv', force=False)

Export File progress: |███████████████████████████████████████████████████| 100%


# Conclusion

In this assignment, I separated into two parts to finish, the first one is data cleaning and preprocessing, and the second part is to use H2O library to make it better. <br><br>

In the first part, I use STEMMER to clean the data, which changes some miss typing words to correct version. Then, I tokenize the words using CountVectorizer. Therefore, we get the mean cross-validation score from Part One, which is about 0.603. Before entering to part two, I save the predictions I get from part one and save them as a new column, this new dataset is the one we are going to use in part two. <br><br>

Instead of importing the original dataset, we will import the one with new columns generated from Part One and then processing them by using H2O Auto ML. However H2O and sklearn cannot exist on the same notebook, that is the reason we have to create two notebooks and finish the work separately. In H2O part, first we initialize H2O and then set up AutoML. AutoML will help us find out the best model fit in our dataset by compare the rmse, log loss and so on from the leaderboard. In our case, the XGBoost run the first place on our leaderboard so we will run this model and let our dataset fit into it. By using XGBoost model, we predict our dataset result in 0.604.  <br><br>
This is a simple combination of sklearn and H2O, however the overfitting problem is a huge problem I should pay more attention to.

# Contributions

By own: 60%<br>
By external source:40%

# Citations

https://www.kaggle.com/sshadylov/sentiment-analysis-on-rotten-tomatoes-dataset<br>
https://github.com/h2oai/h2o-tutorials/blob/master/training/h2o_3_hands_on/nlp/WordEmbeddings-PreRun.ipynb

# License

Copyright 2019 RuisiGu

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

In [ ]:
h2o.cluster().shutdown()